src/MultiFormats.Tests/MultiAddressTest.cs

```diff
@@ -0,0 +1,437 @@
+        public void Parsing()
+        {
+            var a = new MultiAddress(Somewhere);
+            Assert.AreEqual(3, a.Protocols.Count);
+            Assert.AreEqual("ip4", a.Protocols[0].Name);
+            Assert.AreEqual("10.1.10.10", a.Protocols[0].Value);
+            Assert.AreEqual("tcp", a.Protocols[1].Name);
+            Assert.AreEqual("29087", a.Protocols[1].Value);
+            Assert.AreEqual("ipfs", a.Protocols[2].Name);
+            Assert.AreEqual("QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC", a.Protocols[2].Value);
+
+            Assert.AreEqual(0, new MultiAddress((string) null).Protocols.Count);
+            Assert.AreEqual(0, new MultiAddress("").Protocols.Count);
+            Assert.AreEqual(0, new MultiAddress("  ").Protocols.Count);
+        }
+
+        [TestMethod]
+        public void Unknown_Protocol_Name()
+        {
+            ExceptionAssert.Throws<FormatException>(() => new MultiAddress("/foobar/123"));
+        }
+
+        [TestMethod]
+        public void Missing_Protocol_Name() { ExceptionAssert.Throws<FormatException>(() => new MultiAddress("/")); }
+
+        [TestMethod]
+        public new void ToString() { Assert.AreEqual(Somewhere, new MultiAddress(Somewhere).ToString()); }
+
+        [TestMethod]
+        public void Value_Equality()
+        {
+            var a0 = new MultiAddress(Somewhere);
+            var a1 = new MultiAddress(Somewhere);
+            var b = new MultiAddress(Nowhere);
+            MultiAddress c = null;
+            MultiAddress d = null;
+
+            Assert.IsTrue(c == d);
+            Assert.IsFalse(c == b);
+            Assert.IsFalse(b == c);
+
+            Assert.IsFalse(c != d);
+            Assert.IsTrue(c != b);
+            Assert.IsTrue(b != c);
+
+#pragma warning disable 1718
+            Assert.IsTrue(a0 == a0);
+            Assert.IsTrue(a0 == a1);
+            Assert.IsFalse(a0 == b);
+
+#pragma warning disable 1718
+            Assert.IsFalse(a0 != a0);
+            Assert.IsFalse(a0 != a1);
+            Assert.IsTrue(a0 != b);
+
+            Assert.IsTrue(a0.Equals(a0));
+            Assert.IsTrue(a0.Equals(a1));
+            Assert.IsFalse(a0.Equals(b));
+
+            Assert.AreEqual(a0, a0);
+            Assert.AreEqual(a0, a1);
+            Assert.AreNotEqual(a0, b);
+
+            Assert.AreEqual(a0, a0);
+            Assert.AreEqual(a0, a1);
+            Assert.AreNotEqual(a0, b);
+
+            Assert.AreEqual(a0.GetHashCode(), a0.GetHashCode());
+            Assert.AreEqual(a0.GetHashCode(), a1.GetHashCode());
+            Assert.AreNotEqual(a0.GetHashCode(), b.GetHashCode());
+        }
+
+        [TestMethod]
+        public void Bad_Port()
+        {
+            var tcp = new MultiAddress("/tcp/65535");
+            ExceptionAssert.Throws<FormatException>(() => new MultiAddress("/tcp/x"));
+            ExceptionAssert.Throws<FormatException>(() => new MultiAddress("/tcp/65536"));
+
+            var udp = new MultiAddress("/udp/65535");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove the unused local variable 'udp'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)